concurrent_map.go

```diff
@@ -5,75 +5,100 @@ import (
 	"sync"
 )
 
-var SHARD_COUNT = 32
+// ShardCount represent the number of shard the map is devided into.
+var ShardCount = 100
 
-// A "thread" safe map of type string:Anything.
-// To avoid lock bottlenecks this map is dived to several (SHARD_COUNT) map shards.
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
 type ConcurrentMap []*ConcurrentMapShared
 
-// A "thread" safe string to anything map.
+// ConcurrentMapShared is a "thread" safe string to anything map.
 type ConcurrentMapShared struct {
+	mapKeys      []string
+	maxSize      int
 	items        map[string]interface{}
 	sync.RWMutex // Read Write mutex, guards access to internal map.
 }
 
-// Creates a new concurrent map.
-func New() ConcurrentMap {
-	m := make(ConcurrentMap, SHARD_COUNT)
-	for i := 0; i < SHARD_COUNT; i++ {
-		m[i] = &ConcurrentMapShared{items: make(map[string]interface{})}
+// New creates a new concurrent map.
+func New(maxSize int) ConcurrentMap {
+	m := make(ConcurrentMap, ShardCount)
+	shardKeySize := maxSize / ShardCount
+	for i := 0; i < ShardCount; i++ {
+		m[i] = &ConcurrentMapShared{
+			maxSize: shardKeySize,
+			items:   make(map[string]interface{}),
+		}
 	}
 	return m
 }
 
-// GetShard returns shard under given key
+// GetShard returns shard under given key.
 func (m ConcurrentMap) GetShard(key string) *ConcurrentMapShared {
-	return m[uint(fnv32(key))%uint(SHARD_COUNT)]
+	return m[uint(fnv32(key))%uint(ShardCount)]
 }
 
 func (m ConcurrentMap) MSet(data map[string]interface{}) {
 	for key, value := range data {
+		if m.isEvictionOccurred(key) {
+			m.removeFirstElement(key)
+		}
 		shard := m.GetShard(key)
 		shard.Lock()
 		shard.items[key] = value
+		shard.mapKeys = append(shard.mapKeys, key)
 		shard.Unlock()
 	}
 }
 
-// Sets the given value under the specified key.
+// Set sets the given value under the specified key.
 func (m ConcurrentMap) Set(key string, value interface{}) {
 	// Get map shard.
 	shard := m.GetShard(key)
+	if m.isEvictionOccurred(key) {
```

Comment:
Fixed.

Reply:
Not fixed

Reply:
In fact we should not even have to verify if eviction has been occurred here, since we are checking it on the `Put` method from ccache.go line 49. So I completely removed this conditional check.